move to go.mod

sanchezl opened the pull request:
* `go mod init`
* `go mod vendor`

caesarxuchao commented:
Why are there changes in the vendor/ folder? I expect with https://github.com/kubernetes-sigs/kube-storage-version-migrator/pull/41 already merged, this PR will not change the vendor/ folder.

sanchezl commented:
> Why are there changes in the vendor/ folder? I expect with #41 already merged, this PR will not change the vendor/ folder.

`go mod vendor` has different algorithms for deciding what to actually vendor (copies source in packages, not entire modules) and 2) might resolve dependencies differently (but I pinned the dependencies explicitly pinned before).

deads2k commented:
/ok-to-test

caesarxuchao commented on go.mod:
```diff
@@ -0,0 +1,34 @@
+	k8s.io/kube-aggregator v0.0.0-kubernetes-1.16.2
+)
+
+replace (
```
Is the "replace" section manually edited? Why are the specific revisions used here, instead of v0.0.0-kubernetes-1.16.2?

sanchezl replied:
Yes, the `replace` section is manually edited. 

`go.mod` does not allow you to specify a branch or non-semver tag. If you are not using a semver tag (e.g. `v1.0.0`), which kubernetes does not at the moment, you must use a revision, and that is converted to a pseudo-version. Unfortunately, when you run certain `go` commands (such as `go build`), there is a chance that the pseudo version will be replaced with the latest from `master`. So... 
1. I set the `k8s.io/*`  dependencies to a pseudo version of `v0.0.0-kubernetes-1.16.2`. The format accomplishes two things: 1. `go` sees it as a pseudo-version and 2. the `kubernetes-1.16.2` part is a reminder of what I really want to require.
2. Adding `replace` entries for `k8s.io/*` entries accomplishes the following: 1. the psedu-version in the `require` section is ignored, and 2. The revision of the `k8s.io/*` module is effectively _pinned_ and won't change unless explicitly updated.

caesarxuchao replied:
Thanks for the explanation.

If making "v0.0.0-kubernetes-1.16.2" a pseudo version isn't a common practice, can you document the purpose in comments?

In the future, when we need to bump the version of k8s.io/*, I think we need to manually edit the "require" section and the "replace" section, right?

> Unfortunately, when you run certain go commands (such as go build), there is a chance that the pseudo version will be replaced with the latest from master

Can you say what commands exactly?
Anyhow, if we just use the revision based pseudo-version generated by `go`, it's not obvious what tag the revision matches to.

sttts replied:
Let's do the 1.17 port soon. It has proper semver tags.

caesarxuchao commented on go.mod:
```diff
@@ -0,0 +1,34 @@
+module github.com/kubernetes-sigs/kube-storage-version-migrator
```
switch to "sig.k8s.io"?

sanchezl replied:
I will do that in a separate PR.

deads2k replied:
@sanchezl it's a small thing and it's been two months.

deads2k replied:
> @sanchezl it's a small thing and it's been two months.

there's a PR open https://github.com/kubernetes-sigs/kube-storage-version-migrator/pull/46 .

sanchezl commented:
/test pull-kube-storage-version-migrator-disruptive

sanchezl commented:
/retest

caesarxuchao commented:
/retest

sanchezl commented:
/test pull-kube-storage-version-migrator-disruptive

sttts commented:
/lgtm
/approve

deads2k commented:
/approve

sanchezl commented:
/test pull-kube-storage-version-migrator-manually-launched-e2e

sanchezl commented:
/test pull-kube-storage-version-migrator-manually-launched-e2e

sanchezl commented:
/test pull-kube-storage-version-migrator-manually-launched-e2e

sanchezl commented:
/test pull-kube-storage-version-migrator-manually-launched-e2e

caesarxuchao commented:
/test pull-kube-storage-version-migrator-manually-launched-e2e

caesarxuchao commented:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_kube-storage-version-migrator/47/pull-kube-storage-version-migrator-manually-launched-e2e/1220855865574166528

Looks like a real issue:
```
13 migrations haven't succeeded yet
They are:
/admissionregistration.k8s.io.v1.validatingwebhookconfiguragr2xb
/apiregistration.k8s.io.v1.apiservices-h9h8f
/autoscaling.v1.horizontalpodautoscalers-k2gdb
/batch.v1beta1.cronjobs-4n7qd
/coordination.k8s.io.v1.leases-gkzcw
/extensions.v1beta1.ingresses-jsnhf
/rbac.authorization.k8s.io.v1.clusterrolebindings-7xgq2
/rbac.authorization.k8s.io.v1.clusterroles-c86th
/rbac.authorization.k8s.io.v1.rolebindings-thlbc
/rbac.authorization.k8s.io.v1.roles-scfw4
/storage.k8s.io.v1.csinodes-vx4dc
/storage.k8s.io.v1.storageclasses-nfn52
/storage.k8s.io.v1.volumeattachments-dr8vb
retry after 10s
Timed out waiting for migration to complete.
```

caesarxuchao commented:
What's the benefit of keeping the vendor folder?

sanchezl commented:
/test pull-kube-storage-version-migrator-manually-launched-e2e

sanchezl commented:
/test pull-kube-storage-version-migrator-manually-launched-e2e

deads2k commented:
> What's the benefit of keeping the vendor folder?

The vendor folder provides reliable, repeatable builds without vulnerability to the internet.  See the vmware build-farm disaster when a different tag was pushed.  It also allows inspection of all the available code and it prevents failures when libraries completely disappear.

Basically, all the reasons we kept vendor in k/k are still valid and this component should follow suit.

sanchezl commented:
> Looks like a real issue:
> 
> ```
> 13 migrations haven't succeeded yet
> . . . 
> Timed out waiting for migration to complete.
> ```

I've added some logging to the migrator, but the job never seems to capture more than around 10 lines of output. I found the migrator getting stuck when retrying errs that never succeed:
1. when listing resources, if a Not Found (404) error happens here (e.g. bad gvr), migrator will loop here forever (I added an IsNotFound check here to fail the migration).
2. sometimes an error deemed "temporary", never goes away, usually related to permissions, such as:
```
I0128 23:00:56.264015       1 round_trippers.go:423] curl -k -v -XPUT  -H "Accept: application/json" -H "User-Agent: storage-version-migration-migrator/vb04d680" -H "Authorization: Bearer XXX" 'https://10.0.0.1:443/apis/rbac.authorization.k8s.io/v1//edit'
I0128 23:00:56.303078       1 round_trippers.go:443] PUT https://10.0.0.1:443/apis/rbac.authorization.k8s.io/v1/clusterroles/edit 403 Forbidden in 39 milliseconds
I0128 23:00:56.303160       1 round_trippers.go:449] Response Headers:
I0128 23:00:56.303170       1 round_trippers.go:452]     Audit-Id: aed0307a-49ed-4e0b-9514-92ed370669b7
I0128 23:00:56.303176       1 round_trippers.go:452]     Content-Type: application/json
I0128 23:00:56.303182       1 round_trippers.go:452]     Date: Tue, 28 Jan 2020 23:00:56 GMT
I0128 23:00:56.303455       1 request.go:1017] Response Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"clusterroles.rbac.authorization.k8s.io \"edit\" is forbidden: user \"system:serviceaccount:kube-system:default\" (groups=[\"system:serviceaccounts\" \"system:serviceaccounts:kube-system\" \"system:authenticated\"]) is attempting to grant RBAC permissions not currently held: ...
```
I've created a service account with cluster:admin role for migrator to use, just to see it get past #2. I'm not sure why this would be intermittent.

sanchezl commented:
/test pull-kube-storage-version-migrator-disruptive

caesarxuchao commented:
> when listing resources, if a Not Found (404) error happens here (e.g. bad gvr), migrator will loop here forever (I added an IsNotFound check here to fail the migration).

@sanchezl thanks for the investigation. Can you say more about this bug? Is it stuck at this [line](https://github.com/kubernetes-sigs/kube-storage-version-migrator/blob/221b6ef3b41f61395b0ef76c6125cf4a9f1d824d/pkg/migrator/core.go#L97)? Is the storage migrator handling a k8s native resource or a CRD?

sanchezl commented:
> > when listing resources, if a Not Found (404) error happens here (e.g. bad gvr), migrator will loop here forever (I added an IsNotFound check here to fail the migration).
> 
> @sanchezl thanks for the investigation. Can you say more about this bug? Is it stuck at this [line](https://github.com/kubernetes-sigs/kube-storage-version-migrator/blob/221b6ef3b41f61395b0ef76c6125cf4a9f1d824d/pkg/migrator/core.go#L97)? Is the storage migrator handling a k8s native resource or a CRD?

This particular bug does not happen in any of the tests in this repo. I found it when testing on my own clusters and accidentally providing invalid GVRs.

sanchezl commented:
/retest

sanchezl commented:
/retest

sanchezl commented:
@caesarxuchao , @deads2k PTAL,  I've finally gotten the tests to run consistently.

sanchezl commented:
/retest

deads2k commented on pkg/controller/kubemigrator.go:
```diff
@@ -120,11 +121,13 @@ func (km *KubeMigrator) processOne(obj interface{}) error {
 	metrics.Metrics.ObserveFailedMigration(resource(m).String())
-	return err
+	klog.Errorf("%v: migration failed: %v", m.Name, err)
+	return updateErr
```
thi sis a change in behavior. Why shouldn't we return the failure when we failed.  This allows us to return nil after having an error.  If this is correct, add comment explaining why.

sanchezl replied:
Did not intend to change behavior. Any recommendations for handling the `updateErr` ? 
Please comment at https://github.com/kubernetes-sigs/kube-storage-version-migrator/pull/58 instead.

sttts commented:
We import sigs.k8s.io/structured-merge-diff via versioned imports. This breaks the use of dep. We need this now.

caesarxuchao commented:
/lgtm
/approve
